### PR TITLE
Add missing sequence number check and include equivocation/invalid block trackers

### PR DESCRIPTION
**Pull request by KentShikama**
## Overview
Provide a brief description of what this PR does, and why it's needed.

Includes refactoring of block validation. I need some help from @rabbitonweb or @goral09 to get ErrorHandler to work with Casper until them I'm just going to stick with what works/compiles (aka this status pass through thing).

### Does this PR relate to an RChain JIRA issue? 
If applicable, add link to corresponding JIRA issue.

This is a follow up for https://rchain.atlassian.net/browse/RHOL-421

### Complete this checklist before you submit the PR
- [X] This PR contains no more than 200 lines of code, excluding test code.
- [X] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [X] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/Casper.scala [hunk 1]
```diff
@@ -244,6 +259,70 @@ sealed abstract class MultiParentCasperInstances {
           maybeCheckPoint
         }
 
+      private def attemptAdd(b: BlockMessage): F[BlockStatus] =
+        for {
+          status <- validate(b)
+          _      <- addEffects(status, b)
+        } yield status
+
+      // TODO: There must be a way to do this without code duplication
```
[the remark this answers is not in the document]
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala]: Yeah I like this. What do you say @goral09?

**Comment by KentShikama** [comment]
@birchmd PTAL

File: casper/src/main/scala/coop/rchain/casper/Casper.scala [hunk 2]
```diff
@@ -244,6 +238,42 @@ sealed abstract class MultiParentCasperInstances {
...
+       * TODO: Put tuplespace validation back in after we have deterministic unforgeable names.
+       *
+       * We want to catch equivocations only after we confirm that the block completing
+       * the equivocation is otherwise valid.
+       */
+      private def attemptAdd(b: BlockMessage): F[BlockStatus] =
+        for {
+          dag                  <- Capture[F].capture { _blockDag.get }
+          postValidationStatus <- Validate.validateBlockSummary[F](b, genesis, dag)
+          // postTransactionsCheckStatus <- validateTransactions(...)
```
Comment (birchmd) [inline casper/src/main/scala/coop/rchain/casper/Casper.scala]: Please remove old, commented out code

File: casper/src/main/scala/coop/rchain/casper/Casper.scala [hunk 3]
```diff
@@ -244,6 +238,42 @@ sealed abstract class MultiParentCasperInstances {
...
+          postEquivocationCheckStatus <- postValidationStatus.traverse(_ =>
+                                          equivocationsCheck(b, dag, expectedBlockRequests))
+          status = postEquivocationCheckStatus.merge
+          _      <- addEffects(status, b)
+        } yield status
+
+      private def equivocationsCheck(
+          block: BlockMessage,
+          dag: BlockDag,
+          expectedBlockRequests: mutable.HashSet[BlockHash]): F[BlockStatus] = {
```
Comment (birchmd) [inline casper/src/main/scala/coop/rchain/casper/Casper.scala]: You don't use this argument in the function body, why is it here?

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 4]
```diff
@@ -31,15 +28,59 @@ object Validate {
...
+        } yield false
     }
 
-  def blockNumber[F[_]: Applicative: Log](b: BlockMessage, dag: BlockDag): F[Boolean] = {
+  def validateBlockSummary[F[_]: Monad: Log](block: BlockMessage,
+                                             genesis: BlockMessage,
+                                             dag: BlockDag): F[Either[BlockStatus, BlockStatus]] =
+    for {
+      status <- Validate.missingBlocks[F](block, dag)
+      status <- status.traverse(_ => Validate.blockNumber[F](block, dag))
```
Comment (birchmd) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala]: You can reuse the name `status` over and over, and that works? Weird... I don't really like it though and would prefer if each was named differently. E.g. `missingStatus`, `numberStatus`, `parentsStatus`, etc.

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 5]
```diff
@@ -31,15 +28,59 @@ object Validate {
...
+        for {
+          _ <- Log[F].warn(
+                ignore(b, s"block creator ${PrettyPrinter.buildString(b.sender)} has 0 weight."))
+        } yield false
     }
 
-  def blockNumber[F[_]: Applicative: Log](b: BlockMessage, dag: BlockDag): F[Boolean] = {
+  def validateBlockSummary[F[_]: Monad: Log](block: BlockMessage,
+                                             genesis: BlockMessage,
+                                             dag: BlockDag): F[Either[BlockStatus, BlockStatus]] =
```
Comment (birchmd) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala]: I don't love the `Either[BlockStatus, BlockStatus]` return type. Just looking at that alone makes you wonder why use `Either` at all? I still think that taking `Valid` out of the `BlockStatus` trait and renaming the trait to `BlockError` would be better. The return type could then be `Either[BlockError, Valid]` (if you kept around the `Valid` object), or `Either[BlockError, BlockMessage]`. I prefer the latter, but both convey what is going on more clearly.

File: casper/src/main/scala/coop/rchain/casper/Casper.scala (hunk 2 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala]: No this is new code that we need to insert in the future

Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala]: Then maybe a TODO? Can be done in future commit though, not a PR blocker

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala]: The TODO exists above the method

**Pull request merged**